src/main/cpp/src/from_json_to_structs.cu

```diff
@@ -0,0 +1,922 @@
+      });
+
+    auto const out_count_it =
+      thrust::make_zip_iterator(quote_counts.begin(), remove_counts.begin());
+
+    std::size_t temp_storage_bytes = 0;
+    cub::DeviceSegmentedReduce::Reduce(nullptr,
+                                       temp_storage_bytes,
+                                       check_it,
+                                       out_count_it,
+                                       string_count,
+                                       in_offsets,
+                                       in_offsets + 1,
+                                       plus_op,
+                                       count_type{0, 0},
+                                       stream.value());
+    auto d_temp_storage = rmm::device_buffer{temp_storage_bytes, stream};
+    cub::DeviceSegmentedReduce::Reduce(d_temp_storage.data(),
+                                       temp_storage_bytes,
+                                       check_it,
+                                       out_count_it,
+                                       string_count,
+                                       in_offsets,
+                                       in_offsets + 1,
+                                       plus_op,
+                                       count_type{0, 0},
+                                       stream.value());
+  }
+
+  auto const out_size_it = cudf::detail::make_counting_transform_iterator(
+    0,
+    cuda::proclaim_return_type<cudf::size_type>(
+      [offsets       = in_offsets,
+       quote_counts  = quote_counts.begin(),
+       remove_counts = remove_counts.begin()] __device__(auto idx) {
+        auto const input_size = offsets[idx + 1] - offsets[idx];
+        // If the current row is a non-quoted string, just return the original string.
+        if (quote_counts[idx] == 0) { return static_cast<cudf::size_type>(input_size); }
+        // Otherwise, we will modify the string, removing characters '"' and ','.
+        return static_cast<cudf::size_type>(input_size - remove_counts[idx]);
+      }));
+  auto [offsets_column, bytes] = cudf::strings::detail::make_offsets_child_column(
+    out_size_it, out_size_it + string_count, stream, mr);
+
+  // If the output strings column does not change in its total bytes, we know that it does not have
+  // any '"' or ',' characters.
+  if (bytes == input_sv.chars_size(stream)) {
+    return string_to_decimal(precision,
+                             output_type.scale(),
+                             input_sv,
+                             /*ansi_mode*/ false,
+                             /*strip*/ false,
+                             stream,
+                             mr);
+  }
+
+  auto const out_offsets =
+    cudf::detail::offsetalator_factory::make_input_iterator(offsets_column->view());
+  auto chars_data = rmm::device_uvector<char>(bytes, stream, mr);
+
+  // Since the strings store decimal numbers, they should be very short.
```

Comment:
nit: Technically with a DECIMAL_128 it is possible to have 158-bytes per string, but yes, this is generally true.